src/Samples/DSSample.cpp

```diff
@@ -0,0 +1,104 @@
+#include "../../headers/Samples/DSSample.h"
+
+void DSSample::add(const int &value) {
+    this->baseAdd(value);
+}
+
+void DSSample::add(const float &value) {
+    this->baseAdd(value);
+}
+
+void DSSample::add(const double &value) {
+    this->baseAdd(value);
+}
+
+void DSSample::add(const bool &value) {
+    this->baseAdd(value);
+}
+
+void DSSample::add(const long &value) {
+    this->baseAdd(value);
+}
+
+void DSSample::addCategorical(const int &value, const std::vector<int> &possibleValues) {
+    this->baseAddCategorical(value, possibleValues);
+}
+
+void DSSample::addCategorical(const float &value, const std::vector<float> &possibleValues) {
+    this->baseAddCategorical(value, possibleValues);
+}
+
+void DSSample::addCategorical(const double &value, const std::vector<double> &possibleValues) {
+    this->baseAddCategorical(value, possibleValues);
+}
+
+void DSSample::addCategorical(const bool &value, const std::vector<bool> &possibleValues) {
+    this->baseAddCategorical(value, possibleValues);
+}
+
+void DSSample::addCategorical(const long &value, const std::vector<long> &possibleValues) {
+    this->baseAddCategorical(value, possibleValues);
+}
+
+void DSSample::addCategorical(const char &value, const std::vector<char> &possibleValues) {
+    this->baseAddCategorical(value, possibleValues);
+}
+
+void DSSample::reinitData() {
+    this->data.clear();
+}
+
+double* DSSample::getData() {
+    return &this->data[0];
+}
+
+template<class T>
+void DSSample::baseAdd(const T &value) {
+    this->data.emplace_back((double) value);
+}
+
+template<class T>
+void DSSample::baseAddCategorical(const T &value, const std::vector<T> &possibleValues) {
+    for(auto it = possibleValues.begin(); it != possibleValues.end(); ++it)
+    {
+        if(*it == value)
+            this->data.emplace_back((double) 1);
+        else
+            this->data.emplace_back((double) 0);
+    }
+}
+
+template<class T>
+void DSSample::baseAddCategorical(const T &value, const T *possibleValues) {
+    for(auto i = 0; i < sizeof(possibleValues); ++i)
```

Comment:
Try not to use sizeof on arrays when you do not know if they are instantiated dynamically or statically, in general prefer asking for a size in parameter

Reply:
Even worse in this context, ```sizeof``` is evaluating the size of a pointer, not the array content's